compiler/dyno/lib/uast/Builder.cpp

```diff
@@ -280,12 +289,118 @@ void Builder::doAssignIDs(AstNode* ast, UniqueString symbolPath, int& i,
   if (search != notedLocations_.end()) {
     assert(!search->second.isEmpty());
     idToLocation_[ast->id()] = search->second;
+    if (ieNode) {
+      checkConfigPreviouslyUsed(ast, configNameUsed);
+    }
   } else {
     assert(false && "Location for all ast should be set by noteLocation");
   }
 }
 
-AstList Builder::flattenTopLevelBlocks(AstList lst) {
+  void Builder::checkConfigPreviouslyUsed(const AstNode* ast, std::string& configNameUsed) {
+    parsing::useConfigParam(context(), configNameUsed, ast->id());
+    auto usedId = parsing::nameToConfigParamId(context(), configNameUsed);
+
+    if (usedId != ast->id()) {
+      std::string err = "ambiguous config name (";
+      err += configNameUsed;
+      err += ")";
+      std::string otherLoc = "also defined here";
+      std::string disambiguate = "(disambiguate using -s<modulename>.";
+      disambiguate += configNameUsed;
+      disambiguate += "...)";
+      ErrorMessage errorMessage = ErrorMessage(ast->id(), idToLocation_[ast->id()], err, ErrorMessage::ERROR );
+      ErrorMessage noteOtherLoc = ErrorMessage(ast->id(), idToLocation_[usedId], otherLoc, ErrorMessage::NOTE);
+      ErrorMessage noteDisambiguate = ErrorMessage(ast->id(), idToLocation_[usedId], disambiguate, ErrorMessage::NOTE);
+      addError(errorMessage);
+      addError(noteOtherLoc);
+      addError(noteDisambiguate);
+    }
+  }
+
+  std::tuple<AstNode*,std::string> Builder::checkAndUpdateConfig(AstNode* ast, pathVecT& pathVec) {
+    AstNode* ret = nullptr;
```

Comment:
This function mixes a bunch of different concerns. Could you split out two helper functions?

1. A helper function to tell us if the config matched something from the command line, what is the string value of the config from the command line? I would imagine that this helper would work with pathVec etc.
2. A helper function for doing the actual transformation (parsing, extracting the initExpr from the parsed uAST and setting the initExpr in the uAST to keep).

I also had not yet followed the code checking for various error situations but I think that can be simplified with this division (and maybe with other helper functions). You can also add state to the Builder to track which command-line-named-config-vars have been set so far, which might help with the errors.

Reply:
I broke this up into a function to get a command-line config value, one to parse the name/value, one to update the config's initExpr, and one to check if the config was previously assigned/used.